// @assistant-ui/react/dist/legacy-runtime/runtime-cores/external-store/ExternalStoreThreadListRuntimeCore.js
const EMPTY_ARRAY = Object.freeze([]);
const DEFAULT_THREAD_ID = "DEFAULT_THREAD_ID";
const DEFAULT_THREADS = Object.freeze([DEFAULT_THREAD_ID]);
const DEFAULT_THREAD = Object.freeze({
    id: DEFAULT_THREAD_ID,
    remoteId: undefined,
    externalId: undefined,
    status: "regular",
});
const RESOLVED_PROMISE = Promise.resolve();
const DEFAULT_THREAD_DATA = Object.freeze({
    [DEFAULT_THREAD_ID]: DEFAULT_THREAD,
});
export class ExternalStoreThreadListRuntimeCore {
    adapter;
    threadFactory;
    _mainThreadId = DEFAULT_THREAD_ID;
    _threads = DEFAULT_THREADS;
    _archivedThreads = EMPTY_ARRAY;
    _threadData = DEFAULT_THREAD_DATA;
    get isLoading() {
        return this.adapter.isLoading ?? false;
    }
    get newThreadId() {
        return undefined;
    }
    get threadIds() {
        return this._threads;
    }
    get archivedThreadIds() {
        return this._archivedThreads;
    }
    get threadData() {
        return this._threadData;
    }
    getLoadThreadsPromise() {
        return RESOLVED_PROMISE;
    }
    _mainThread;
    get mainThreadId() {
        return this._mainThreadId;
    }
    constructor(adapter = {}, threadFactory) {
        this.adapter = adapter;
        this.threadFactory = threadFactory;
        this._mainThread = this.threadFactory();
        this.__internal_setAdapter(adapter, true);
    }
    getMainThreadRuntimeCore() {
        return this._mainThread;
    }
    getThreadRuntimeCore() {
        throw new Error("Method not implemented.");
    }
    getItemById(threadId) {
        for (const thread of this.adapter.threads ?? []) {
            if (thread.id === threadId)
                return thread;
        }
        for (const thread of this.adapter.archivedThreads ?? []) {
            if (thread.id === threadId)
                return thread;
        }
        if (threadId === DEFAULT_THREAD_ID)
            return DEFAULT_THREAD;
        return undefined;
    }
    __internal_setAdapter(adapter, initialLoad = false) {
        const previousAdapter = this.adapter;
        this.adapter = adapter;
        const newThreadId = adapter.threadId ?? DEFAULT_THREAD_ID;
        const newThreads = adapter.threads ?? EMPTY_ARRAY;
        const newArchivedThreads = adapter.archivedThreads ?? EMPTY_ARRAY;
        const previousThreadId = previousAdapter.threadId ?? DEFAULT_THREAD_ID;
        const previousThreads = previousAdapter.threads ?? EMPTY_ARRAY;
        const previousArchivedThreads = previousAdapter.archivedThreads ?? EMPTY_ARRAY;
        if (!initialLoad &&
            previousThreadId === newThreadId &&
            previousThreads === newThreads &&
            previousArchivedThreads === newArchivedThreads) {
            return;
        }
        this._threadData = {
            ...DEFAULT_THREAD_DATA,
            ...Object.fromEntries(adapter.threads?.map((t) => [
                t.id,
                {
                    ...t,
                    remoteId: t.remoteId,
                    externalId: t.externalId,
                    status: "regular",
                },
            ]) ?? []),
            ...Object.fromEntries(adapter.archivedThreads?.map((t) => [
                t.id,
                {
                    ...t,
                    remoteId: t.remoteId,
                    externalId: t.externalId,
                    status: "archived",
                },
            ]) ?? []),
        };
        if (previousThreads !== newThreads) {
            this._threads = this.adapter.threads?.map((t) => t.id) ?? EMPTY_ARRAY;
        }
        if (previousArchivedThreads !== newArchivedThreads) {
            this._archivedThreads =
                this.adapter.archivedThreads?.map((t) => t.id) ?? EMPTY_ARRAY;
        }
        if (previousThreadId !== newThreadId) {
            this._mainThreadId = newThreadId;
            this._mainThread = this.threadFactory();
        }
        this._notifySubscribers();
    }
    async switchToThread(threadId) {
        if (this._mainThreadId === threadId)
            return;
        const onSwitchToThread = this.adapter.onSwitchToThread;
        if (!onSwitchToThread)
            throw new Error("External store adapter does not support switching to thread");
        onSwitchToThread(threadId);
    }
    async switchToNewThread() {
        const onSwitchToNewThread = this.adapter.onSwitchToNewThread;
        if (!onSwitchToNewThread)
            throw new Error("External store adapter does not support switching to new thread");
        onSwitchToNewThread();
    }
    async rename(threadId, newTitle) {
        const onRename = this.adapter.onRename;
        if (!onRename)
            throw new Error("External store adapter does not support renaming");
        onRename(threadId, newTitle);
    }
    async detach() {
        // no-op
    }
    async archive(threadId) {
        const onArchive = this.adapter.onArchive;
        if (!onArchive)
            throw new Error("External store adapter does not support archiving");
        onArchive(threadId);
    }
    async unarchive(threadId) {
        const onUnarchive = this.adapter.onUnarchive;
        if (!onUnarchive)
            throw new Error("External store adapter does not support unarchiving");
        onUnarchive(threadId);
    }
    async delete(threadId) {
        const onDelete = this.adapter.onDelete;
        if (!onDelete)
            throw new Error("External store adapter does not support deleting");
        onDelete(threadId);
    }
    initialize(threadId) {
        return Promise.resolve({ remoteId: threadId, externalId: undefined });
    }
    generateTitle() {
        throw new Error("Method not implemented.");
    }
    _subscriptions = new Set();
    subscribe(callback) {
        this._subscriptions.add(callback);
        return () => this._subscriptions.delete(callback);
    }
    _notifySubscribers() {
        for (const callback of this._subscriptions)
            callback();
    }
}
