// assistant-stream/dist/core/tool/schema-utils.js
function isStandardSchema(schema) {
    return (typeof schema === "object" &&
        schema !== null &&
        "~standard" in schema &&
        typeof schema["~standard"] === "object");
}
function hasToJSONSchemaMethod(schema) {
    return (typeof schema === "object" &&
        schema !== null &&
        "toJSONSchema" in schema &&
        typeof schema.toJSONSchema === "function");
}
function hasToJSONMethod(schema) {
    return (typeof schema === "object" &&
        schema !== null &&
        "toJSON" in schema &&
        typeof schema.toJSON === "function");
}
/**
 * Converts a schema to JSONSchema7.
 * Supports:
 * - StandardSchemaV1 with ~standard.toJSONSchema (e.g., Zod v4)
 * - Objects with toJSONSchema() method
 * - Objects with toJSON() method
 * - Plain JSONSchema7 objects
 */
export function toJSONSchema(schema) {
    // StandardSchemaV1 with ~standard.toJSONSchema (e.g., Zod v4)
    if (isStandardSchema(schema)) {
        const toJSONSchemaMethod = schema["~standard"].toJSONSchema;
        if (typeof toJSONSchemaMethod === "function") {
            return toJSONSchemaMethod();
        }
    }
    // toJSONSchema method on the schema itself
    if (hasToJSONSchemaMethod(schema)) {
        return schema.toJSONSchema();
    }
    // toJSON method on the schema
    if (hasToJSONMethod(schema)) {
        return schema.toJSON();
    }
    // Already a plain JSONSchema7
    return schema;
}
function defaultToolFilter(_name, tool) {
    return !tool.disabled && tool.type !== "backend";
}
/**
 * Converts a record of tools to a record of tool definitions with JSON Schema parameters.
 * By default, filters out disabled tools and backend tools.
 */
export function toToolsJSONSchema(tools, options = {}) {
    if (!tools)
        return {};
    const filter = options.filter ?? defaultToolFilter;
    return Object.fromEntries(Object.entries(tools)
        .filter(([name, tool]) => filter(name, tool) && tool.parameters)
        .map(([name, tool]) => [
        name,
        {
            ...(tool.description && { description: tool.description }),
            parameters: toJSONSchema(tool.parameters),
        },
    ]));
}
